core/src/main/scala/org/bitcoins/core/util/NumberUtil.scala

```diff
@@ -175,19 +177,121 @@ trait NumberUtil extends BitcoinSLogger {
 
     val mantissa =
       new BigInteger(signum, mantissaBytes.toArray)
-
     //guards against a negative exponent, in which case we just shift right
     //see bitcoin core implementation
-    if (significand <= 3) {
-      mantissa.shiftRight(8 * (3 - significand))
+    val result = {
+      if (significand <= 3) {
+
+        val exp = 8 * (3 - significand)
+        //avoid shift right, because of weird behavior on the jvm
+        //https://stackoverflow.com/questions/47519140/bitwise-shift-right-with-long-not-equaling-zero/47519728#47519728
+        mantissa.divide(NumberUtil.pow2(exp).bigInteger)
+      } else {
+        val exponent = significand - 3
+
+        val pow256 = BigInteger.valueOf(256).pow(exponent)
+        mantissa.multiply(pow256)
+      }
+    }
+
+    val nWordNotZero = mantissa != (BigInteger.ZERO)
+    val nWord = BigInt(mantissa)
+    val nSize = nBits.toBigInt / (NumberUtil.pow2(24))
+    val isNegative: Boolean = {
+      //*pfNegative = nWord != 0 && (nCompact & 0x00800000) != 0;
+      nWordNotZero &&
+      (nBits & UInt32.fromHex("00800000")) != UInt32.zero
+    }
+
+    val isOverflow: Boolean = {
+      //nWord != 0 && ((nSize > 34) ||
+      //  (nWord > 0xff && nSize > 33) ||
+      //  (nWord > 0xffff && nSize > 32));
+
+      nWordNotZero && ((nSize > 34) ||
+      (nWord > UInt8.max.toBigInt && nSize > 33) ||
+      (nWord > UInt32.fromHex("ffff").toBigInt && nSize > 32))
+    }
+
+    BlockHeader.TargetDifficultyHelper(result.abs(), isNegative, isOverflow)
+  }
+
+  /**
+    * Compressed the big integer to be used inside of [[org.bitcoins.core.protocol.blockchain.BlockHeader.nBits]]
+    * [[https://github.com/bitcoin/bitcoin/blob/2068f089c8b7b90eb4557d3f67ea0f0ed2059a23/src/arith_uint256.cpp#L226 bitcoin core implementation]]
+    * @param bigInteger
+    * @return
+    */
+  def targetCompression(bigInteger: BigInteger, isNegative: Boolean): UInt32 = {
+    val bytes = bigInteger.toByteArray
+    val bitVec = BitVector(bytes)
+
+    val negativeFlag = UInt32.fromHex("00800000")
+
+    //emulates bits() in arith_uin256.h
+    //Returns the position of the highest bit set plus one, or zero if the
+    //value is zero.
+    //https://github.com/bitcoin/bitcoin/blob/2068f089c8b7b90eb4557d3f67ea0f0ed2059a23/src/arith_uint256.h#L241
+    var size: Int = if (bigInteger == BigInteger.valueOf(0)) {
+      0
     } else {
-      val exponent = significand - 3
+      ((bitVec.length + 7) / 8).toInt
+    }
+
+    var compact: UInt32 = {
+      if (size <= 3) {
+        // GetLow64() << 8 * (3 - nSize);
+        //note: counter intuitively, the expression
+        // 8 * (3 - nSize)
+        //gets evaluated before we apply the shift left operator
+        //verified this property with g++
+        val shiftAmount = 8 * (3 - size)
 
-      val pow256 = BigInteger.valueOf(256).pow(exponent)
-      mantissa.multiply(pow256)
+        val u64 = new BigInteger(1, bitVec.takeRight(64).toByteArray)
```

Comment:
`BigInt`? Also feel like I'm seeing a lot of these weird big integer constructors. Should we create an abstraction/helper method for them?

Reply:
that is a great idea.